### DISCUSSION TITLE
[Android] Bone transformation, odd bone weight?

### DISCUSSION THREAD

**Discussion by wes-nz** [google/filament#4728] [Q&A] [answered]
I am trying to get a skeleton to apply the correct bone weight to the mesh. So far my results have been quite different from other engines. I'd love to get a second opinion in case I am doing something wrong.

Expected result (Blender, SceneKit):

https://user-images.githubusercontent.com/2945380/137070559-b3c3bb54-6806-4e5b-a84a-1ed6e9e1450b.mov

Current result in Filament:

https://user-images.githubusercontent.com/2945380/137070614-6c4ae983-be6f-4740-bfb1-12a9a3b7e615.mov

```Kotlin
modelViewer?.asset?.getFirstEntityByName("NeckBone")?.let { entity ->
    val neckBoneInstance = engine.transformManager.getInstance(entity)
    val transform = defaultNeckTransform!! * rotation(Float3(1f, 0.0f, 0.0f), boneRotationPositionX)
    engine.transformManager.setTransform(neckBoneInstance, transform.toFloatArray())
    modelViewer?.animator?.updateBoneMatrices()
}
```

Alternatively if I try to apply the transform directly to the bone as [explained here](https://github.com/google/filament/discussions/4241) the bone collapses, I assume the transform is reset.

https://user-images.githubusercontent.com/2945380/137070944-f1948520-8280-4c8c-a871-b71c651b98ab.mov

```Kotlin
modelViewer?.asset?.getFirstEntityByName("NeckBone")?.let { entity ->
    val neckBoneInstance = engine.transformManager.getInstance(entity)
    val transform = rotation(Float3(1.0f,0.0f,0.0f), boneRotationPositionX)
    engine.transformManager.setTransform(neckBoneInstance, transpose(transform).toFloatArray())
    modelViewer?.animator?.updateBoneMatrices()
}
```

Any idea would be appreciated, thanks!

**Comment by romainguy** [comment, marked as answer]
You are not setting bone weights here, you are setting a transform on an entity. That's quite different. Bones are modified via the RenderableManager. Which are you trying to do?

**Reply by wes-nz** [reply]
Oh right, of course that makes sense.  The result is much better but as you can see it's still different, the vertices are stretching instead of collapsing. Is there something else I am doing wrong? 

https://user-images.githubusercontent.com/2945380/137206941-ed0f2376-21b1-4e1c-8748-b0a0ff1c8252.mov


![crease](https://user-images.githubusercontent.com/2945380/137208115-d07b27d5-d9cf-4df0-9360-f56b46bc1f3b.jpg)


[Updated project](https://www.dropbox.com/s/6ydpddbpbkgz8a0/Filament%20-%20Bones%20Weight.zip?dl=0)
[Blender file](https://www.dropbox.com/s/1xvr69kr4b378ws/skeleton-bone-weight.blend.zip?dl=0)

**Reply by romainguy** [reply]
Looks like some of the weights are missing/incorrect? Ut looks like some of the vertices are not being influenced.

@prideout is our skinning expert.

**Reply by wes-nz** [reply]
Thanks, I'll wait to hear from him.

On the same topic I would be curious to know if Filament supports multiple entities within one armature. Rotating the bone rotates both the sphere and the torus in other engines. Not sure if this is a Filament limitation or Gltf.

![torus](https://user-images.githubusercontent.com/2945380/137230637-28a6622d-d64c-4577-a3ee-a724c1096ad2.jpg)

**Reply by prideout** [reply]
You might want to try playing with the `normalizeSkinningWeights` field in ModelViewer, this is used in `init` to determine whether or not our loader will force skinning weights to add up to 1.0.

Re: armature, I would need to study your glTF file.  I'd like to understand this better if we behave differently from what you've seen in other engines.

**Reply by wes-nz** [reply]
Sorry @prideout I don't have the file anymore. It's been almost a year, I had to rebuild my game with a different engine.

**Comment by romainguy** [comment]
It could be an issue with the glTF exporter of Blender. Bones could be shared across entities.
